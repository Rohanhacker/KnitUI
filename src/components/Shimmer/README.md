# Shimmer

## Usage

```javascript
import { Shimmer } from "KnitUI"

return <Shimmer height="medium" />
```

| Prop name | Type    | Optional | Default |
| --------- | ------- | -------- | ------- |
| height    | "large" | "medium" | True    | "small" |
