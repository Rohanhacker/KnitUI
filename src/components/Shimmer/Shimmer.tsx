import React from "react"
import styled, { keyframes } from "styled-components"
import { BaseComponentProps } from "../../common/types"

type heightType = "small" | "medium" | "large"

export interface IShimmerProps extends BaseComponentProps {
  /** This is a placeholder description */
  height?: heightType
}

const shim = keyframes`
from {
  left: -100%;
}
to {
  left: 100%;
}
`

function getHeight(height: heightType): number {
  switch (height) {
    case "large":
      return 3.2
    case "medium":
      return 2.8
    default:
      return 2
  }
}

const ShimmerContainer: any = styled.div<{
  customProps: { height: heightType }
}>`
  width: 100%;
  height: ${({ customProps }) => `${getHeight(customProps.height)}rem`};
  background: ${({ theme }) => theme.knitui.chromaPalette.Neutral10};
  border-radius: 4px;
  position: relative;
  overflow: hidden;
`

const ShimmerElement = styled.div`
  width: 50%;
  max-width: 35rem;
  background: ${({ theme }) => theme.knitui.chromaPalette.Neutral40};
  height: 100%;
  border-radius: 4px;
  filter: blur(34px);
  animation: ${shim} 2000ms cubic-bezier(0.5, 0, 0, 1) infinite;
  position: absolute;
`

export default function Shimmer({
  height = "small",
  className,
  style,
}: IShimmerProps) {
  return (
    <ShimmerContainer
      className={className}
      style={style}
      customProps={{ height }}>
      <ShimmerElement />
    </ShimmerContainer>
  )
}
