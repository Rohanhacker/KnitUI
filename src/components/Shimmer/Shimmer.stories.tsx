import React from "react"
import { storiesOf } from "@storybook/react"
import { withKnobs, select } from "@storybook/addon-knobs"
import Shimmer from "./index"

const heightOptions = ["small", "medium", "large"]

storiesOf("Shimmer", module)
  .addDecorator(withKnobs)
  .add("basic", () => (
    <div
      style={{
        width: 420,
      }}>
      <Shimmer height={select("Size", heightOptions, "medium")} />
    </div>
  ))
  .add("multiple shimmers", () => (
    <div
      style={{
        width: 420,
      }}>
      <Shimmer
        style={{
          opacity: 1,
          margin: "14px 0",
        }}
        height={select("Size", heightOptions, "medium")}
      />
      <Shimmer
        style={{
          opacity: 0.75,
          margin: "14px 0",
        }}
        height={select("Size", heightOptions, "medium")}
      />{" "}
      <Shimmer
        style={{
          opacity: 0.5,
          margin: "14px 0",
        }}
        height={select("Size", heightOptions, "medium")}
      />{" "}
      <Shimmer
        style={{
          opacity: 0.25,
          margin: "14px 0",
        }}
        height={select("Size", heightOptions, "medium")}
      />
    </div>
  ))
