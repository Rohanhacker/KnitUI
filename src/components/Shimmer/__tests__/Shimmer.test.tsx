import React from "react"
import Shimmer from "../Shimmer"
import { render, cleanup } from "../../../common/TestUtil"
import { ThemeProvider } from "../../../common/styles"
import "jest-styled-components"
import "jest-dom/extend-expect"

afterEach(cleanup)

describe("Shimmer", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Shimmer />
      </ThemeProvider>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
