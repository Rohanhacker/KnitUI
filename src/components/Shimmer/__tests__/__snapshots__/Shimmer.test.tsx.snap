// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shimmer should render correctly 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  height: 2rem;
  background: #f2f2f2;
  border-radius: 4px;
  position: relative;
  overflow: hidden;
}

.c1 {
  width: 50%;
  max-width: 35rem;
  background: #cccccc;
  height: 100%;
  border-radius: 4px;
  -webkit-filter: blur(34px);
  filter: blur(34px);
  -webkit-animation: gVDJEF 2000ms cubic-bezier(0.5,0,0,1) infinite;
  animation: gVDJEF 2000ms cubic-bezier(0.5,0,0,1) infinite;
  position: absolute;
}

<div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</DocumentFragment>
`;
